### PR TITLE
モデルサーバーの追加

### PR DESCRIPTION
**Pull request by littlemex**
Model Serverを追加した

フィードバックで得たものはIssue化する。
一旦通しでdockerが動いたこのバージョンでv0.0.2のタグを切る。

- [x] baseイメージの追加、buildx bake baseで実行する
- [x] traceイメージの追加、buildx bake traceで実行する 、ptファイルの生成
- [x] model serverイメージの追加、buildx bake modelで実行する、ptファイルを取り込んでAPIサーバを立てる 
- [x] ECRレポジトリは既に存在する前提でpushするスクリプトを準備
- [ ] 現状の作りだとtraceが修正しづらいのでNotebookから修正できるようにする, S3にptを配置する -> Issue化
- [x] Deep Learning用のコンテナを利用する方が簡潔に作れる -> Issue化
- [ ] ベンチマークの仕組みを入れる -> Issue化

使い方

```bash
docker buildx bake base && docker buildx bake trace && docker  buildx bake model
./run-model-server.sh
```

私は[MASK]へ行く、という感じでmasked_indexで指定した部分がマスクされる。

```bash
# リクエストのテスト
curl -X 'POST' \
  'http://localhost/inferences' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "text": "私は東京へ行く",
  "mask_index": 3
}'
```

app/main.pyがmodel serverのコード

```
.
├── app
│   ├── main.py
│   ├── requirements.txt
│   └── run.sh
```

### CODE REVIEW

File: model-server/build/Dockerfile-base [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#Install Neuron PyTorch
+RUN pip3 install torch-neuron neuron-cc[tensorflow] torchvision --extra-index-url=https://pip.repos.neuron.amazonaws.com
+
+RUN pip3 install transformers fugashi ipadic
+
+# For Trace ---------------------------------------
+FROM inf1-base as trace-stage
+
+ENV AWS_NEURON_VISIBLE_DEVICES=ALL
```
Comment (littlemex) [inline model-server/build/Dockerfile-base]: model-server側には設定が効いていないので、base側に設定を移動する

File: model-server/app/main.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    def _conv_to_tokens_from_(self, index):
+        print("item", index.item())
+        print(self.tokenizer.convert_ids_to_tokens(index.item()))
+        return self.tokenizer.convert_ids_to_tokens([index.item()])[0]
+
+    def infer(self, message):
+
+        text = message.text
+        masked_index = message.masked_index
```
Comment (littlemex) [inline model-server/app/main.py]: masked_indexの名前が変な気もするので修正検討する。mask_indexの方がいいのか

File: model-server/app/main.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import os
```
Comment (littlemex) [inline model-server/app/main.py:1]: isortで並び順綺麗にする

File: model-server/app/main.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+path_prefix = "/app/server/models/"
+model_path = os.path.join(path_prefix, "transformers_neuron.pt")
+
+LENGTH = 512
+
+
+class ModelInference:
+    def __init__(self):
+        pretrained = "cl-tohoku/bert-base-japanese-whole-word-masking"
```
Comment (littlemex) [inline model-server/app/main.py]: 勝手にcl-tohokuのやつ使っちゃってるけど、CCのライセンス的に大丈夫なのか調べる

File: model-server/app/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+fastapi
```
Comment (littlemex) [inline model-server/app/requirements.txt:1]: 消すか・・、Prodではfastapi使わないし

File: model-server/trace/tracer.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import torch
+import torch_neuron
+from transformers import BertForMaskedLM, BertJapaneseTokenizer
+
+model_name = "cl-tohoku/bert-base-japanese-whole-word-masking"
+ptname = "transformers_neuron.pt"
+processor = "inf1"
+pipeline_cores = 1
```
Comment (littlemex) [inline model-server/trace/tracer.py]: 使ってない変数を消す

File: model-server/app/main.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    def infer(self, message):
+
+        text = message.text
+        masked_index = message.masked_index
+
+        logger.info(f"Input text : {text}")
+        tokenized_text = self.tokenizer.tokenize(text)
+        logger.info("Tokenized text : " + ",".join(tokenized_text))
+        tokenized_text[masked_index] = "[MASK]"
```
Comment (littlemex) [inline model-server/app/main.py]: この辺はprint debugのためだけのコードなので消すか

File: model-server/build/Dockerfile-base [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                aws-neuron-dkms aws-neuron-tools \
+                -y
+
+ENV PATH=/opt/aws/neuron/bin:$PATH
+
+RUN yum install -y python3.7 python3.7-devel gcc-c++ \
+                   tar gzip ca-certificates procps net-tools which vim wget libgomp htop jq bind-utils bc
+
+#Install Neuron PyTorch
+RUN pip3 install torch-neuron neuron-cc[tensorflow] torchvision --extra-index-url=https://pip.repos.neuron.amazonaws.com
```
Comment (littlemex) [inline model-server/build/Dockerfile-base]: pip install torch-neuron neuron-cc[tensorflow] "protobuf<4" torchvision

File: model-server/build/add-neuron-repo.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
Comment (littlemex) [inline model-server/build/add-neuron-repo.sh]: いらんな
